File: backend/src/monitoring/analyzers/connection_analyzer.py
import threading
import time
from collections import defaultdict
from typing import List, Dict, Optional, Callable, Any

# Data structures

class ConnectionData:
    def __init__(self, ip: str, timestamp: float, type: str, resource=None, userAgent=None, bytesTransferred=0, duration=None):
        self.ip = ip
        self.timestamp = timestamp
        self.type = type
        self.resource = resource
        self.userAgent = userAgent
        self.bytesTransferred = bytesTransferred
        self.duration = duration

class SuspiciousActivity:
    def __init__(self, ip: str, type: str, timestamp: float, details: str, severity: str):
        self.ip = ip
        self.type = type
        self.timestamp = timestamp
        self.details = details
        self.severity = severity

class ConnectionMetrics:
    def __init__(self):
        self.totalConnections = 0
        self.activeConnections = 0
        self.connectionsByType = {}
        self.averageConnectionDuration = 0
        self.totalBytesTransferred = 0
        self.uniqueIPs = 0
        self.suspiciousActivity: List[SuspiciousActivity] = []

class ConnectionAnalyzer:
    def __init__(self):
        self.connections: Dict[str, List[ConnectionData]] = defaultdict(list)
        self.activeConnections: set = set()
        self.metrics = ConnectionMetrics()
        self.maxConnectionHistory = 1000
        self.suspiciousThresholds = {
            'rapidRequests': 10,
            'largeDownload': 100 * 1024 * 1024,
            'connectionDuration': 300000  # 5 minutes in ms
        }
        self.suspicious_callbacks: List[Callable[[SuspiciousActivity], Any]] = []
        self.analysisInterval: Optional[threading.Timer] = None
        self._start_analysis()

    def initialize_metrics(self):
        return ConnectionMetrics()

    def log_connection(self, ip: str, type: str, resource=None, userAgent=None) -> str:
        connectionId = self._generate_connection_id()
        connectionData = ConnectionData(
            ip=ip,
            timestamp=time.time() * 1000,
            type=type,
            resource=resource,
            userAgent=userAgent
        )
        self.connections[ip].append(connectionData)
        if len(self.connections[ip]) > self.maxConnectionHistory:
            self.connections[ip].pop(0)
        self.activeConnections.add(connectionId)
        self._update_metrics()
        self._check_suspicious_activity(ip, connectionData)
        return connectionId

    def update_connection_bytes(self, connectionId: str, bytes_: int):
        for ip, connections in self.connections.items():
            for connection in connections:
                if self._generate_connection_id_from(connection) == connectionId:
                    connection.bytesTransferred = bytes_
                    self._update_metrics()
                    if bytes_ > self.suspiciousThresholds['largeDownload']:
                        self._flag_suspicious_activity(ip, 'large_download',
                            f"Large download detected: {self._format_bytes(bytes_)}", 'medium')
                    return

    def close_connection(self, connectionId: str):
        self.activeConnections.discard(connectionId)
        for connections in self.connections.values():
            for connection in connections:
                if self._generate_connection_id_from(connection) == connectionId:
                    connection.duration = time.time() * 1000 - connection.timestamp
                    break
        self._update_metrics()

    def _generate_connection_id(self):
        return f"{int(time.time() * 1000)}_{hex(int(time.time() * 1000000) % 0xffffff)[2:]}"

    def _generate_connection_id_from(self, connection: ConnectionData):
        return f"{connection.ip}_{int(connection.timestamp)}_{connection.type}"

    def _update_metrics(self):
        allConnections = []
        uniqueIPs = set()
        for ip, connections in self.connections.items():
            uniqueIPs.add(ip)
            allConnections.extend(connections)
        self.metrics.totalConnections = len(allConnections)
        self.metrics.activeConnections = len(self.activeConnections)
        self.metrics.uniqueIPs = len(uniqueIPs)
        # Connection types
        self.metrics.connectionsByType = {}
        for connection in allConnections:
            self.metrics.connectionsByType[connection.type] = self.metrics.connectionsByType.get(connection.type, 0) + 1
        # Average duration and total bytes
        completedConnections = [c for c in allConnections if c.duration]
        if completedConnections:
            self.metrics.averageConnectionDuration = sum(c.duration for c in completedConnections) / len(completedConnections)
        else:
            self.metrics.averageConnectionDuration = 0
        self.metrics.totalBytesTransferred = sum(c.bytesTransferred or 0 for c in allConnections)

    def _check_suspicious_activity(self, ip: str, newConnection: ConnectionData):
        ipConnections = self.connections.get(ip, [])
        now = time.time() * 1000
        recentConnections = [c for c in ipConnections if now - c.timestamp < 60000]
        # rapid requests
        if len(recentConnections) >= self.suspiciousThresholds['rapidRequests']:
            self._flag_suspicious_activity(ip, 'rapid_requests',
                f"{len(recentConnections)} requests in the last minute", 'high')
        # repeated file downloads
        typeCount = sum(1 for c in recentConnections if c.type == newConnection.type)
        if typeCount > 5 and newConnection.type == 'file_download':
            self._flag_suspicious_activity(ip, 'unusual_pattern',
                f"Repeated {newConnection.type} requests", 'medium')

    def _flag_suspicious_activity(self, ip: str, type_: str, details: str, severity: str):
        activity = SuspiciousActivity(
            ip=ip,
            type=type_,
            timestamp=time.time() * 1000,
            details=details,
            severity=severity
        )
        self.metrics.suspiciousActivity.append(activity)
        if len(self.metrics.suspiciousActivity) > 100:
            self.metrics.suspiciousActivity.pop(0)
        # Emit event via callback
        for cb in self.suspicious_callbacks:
            cb(activity)

    def add_suspicious_callback(self, callback: Callable[[SuspiciousActivity], Any]):
        self.suspicious_callbacks.append(callback)

    def _start_analysis(self):
        def periodic():
            self._cleanup_old_connections()
            self._perform_security_analysis()
            self.analysisInterval = threading.Timer(30, periodic)
            self.analysisInterval.daemon = True
            self.analysisInterval.start()
        periodic()

    def _cleanup_old_connections(self):
        cutoffTime = time.time() * 1000 - 24 * 60 * 60 * 1000
        for ip in list(self.connections.keys()):
            filtered = [c for c in self.connections[ip] if c.timestamp > cutoffTime]
            if not filtered:
                del self.connections[ip]
            else:
                self.connections[ip] = filtered
        self.metrics.suspiciousActivity = [
            a for a in self.metrics.suspiciousActivity if time.time() * 1000 - a.timestamp < 24 * 60 * 60 * 1000
        ]

    def _perform_security_analysis(self):
        now = time.time() * 1000
        for ip, connections in self.connections.items():
            recentConnections = [c for c in connections if now - c.timestamp < 300000]
            # potential DDoS
            if len(recentConnections) > 20:
                self._flag_suspicious_activity(ip, 'unusual_pattern',
                    'Potential DDoS pattern detected', 'high')
            # data exfiltration
            totalBytes = sum(c.bytesTransferred or 0 for c in recentConnections)
            if totalBytes > 500 * 1024 * 1024:
                self._flag_suspicious_activity(ip, 'large_download',
                    'Potential data exfiltration detected', 'high')

    def _format_bytes(self, bytes_):
        if bytes_ >= 1024 ** 3:
            return f"{bytes_ / 1024 ** 3:.2f} GB"
        elif bytes_ >= 1024 ** 2:
            return f"{bytes_ / 1024 ** 2:.2f} MB"
        elif bytes_ >= 1024:
            return f"{bytes_ / 1024:.2f} KB"
        return f"{bytes_} B"

    def get_metrics(self):
        return self.metrics

    def get_connection_history(self, ip=None):
        if ip:
            return self.connections.get(ip, [])
        allConnections = []
        for connections in self.connections.values():
            allConnections.extend(connections)
        return sorted(allConnections, key=lambda c: -c.timestamp)

    def destroy(self):
        if self.analysisInterval:
            self.analysisInterval.cancel()
            self.analysisInterval = None
        self.connections.clear()
        self.activeConnections.clear()
        self.suspicious_callbacks.clear()

# Example usage (no Flask endpoints here yet):
# analyzer = ConnectionAnalyzer()
# connection_id = analyzer.log_connection('127.0.0.1', 'file_download')
# analyzer.update_connection_bytes(connection_id, 150000000)
# analyzer.close_connection(connection_id)
# print(analyzer.get_metrics().__dict__)

File: backend/src/monitoring/analyzers/ConnectionAnalyzer.ts
import { EventEmitter } from 'events';

interface ConnectionData {
  ip: string;
  timestamp: number;
  type: 'file_download' | 'stream_request' | 'websocket';
  resource?: string;
  userAgent?: string;
  bytesTransferred?: number;
  duration?: number;
}

interface ConnectionMetrics {
  totalConnections: number;
  activeConnections: number;
  connectionsByType: Record<string, number>;
  averageConnectionDuration: number;
  totalBytesTransferred: number;
  uniqueIPs: number;
  suspiciousActivity: SuspiciousActivity[];
}

interface SuspiciousActivity {
  ip: string;
  type: 'rapid_requests' | 'large_download' | 'unusual_pattern';
  timestamp: number;
  details: string;
  severity: 'low' | 'medium' | 'high';
}

export class ConnectionAnalyzer extends EventEmitter {
  private connections: Map<string, ConnectionData[]> = new Map();
  private activeConnections: Set<string> = new Set();
  private metrics: ConnectionMetrics;
  private analysisInterval: NodeJS.Timeout | null = null;
  private readonly maxConnectionHistory = 1000;
  private readonly suspiciousThresholds = {
    rapidRequests: 10, // requests per minute
    largeDownload: 100 * 1024 * 1024, // 100MB
    connectionDuration: 300000 // 5 minutes
  };

  constructor() {
    super();
    this.metrics = this.initializeMetrics();
    this.startAnalysis();
  }

  private initializeMetrics(): ConnectionMetrics {
    return {
      totalConnections: 0,
      activeConnections: 0,
      connectionsByType: {},
      averageConnectionDuration: 0,
      totalBytesTransferred: 0,
      uniqueIPs: 0,
      suspiciousActivity: []
    };
  }

  public logConnection(
    ip: string, 
    type: ConnectionData['type'], 
    resource?: string,
    userAgent?: string
  ): string {
    const connectionId = this.generateConnectionId();
    const connectionData: ConnectionData = {
      ip,
      timestamp: Date.now(),
      type,
      resource,
      userAgent,
      bytesTransferred: 0
    };

    // Store connection data
    if (!this.connections.has(ip)) {
      this.connections.set(ip, []);
    }
    
    const ipConnections = this.connections.get(ip)!;
    ipConnections.push(connectionData);
    
    // Limit history per IP
    if (ipConnections.length > this.maxConnectionHistory) {
      ipConnections.shift();
    }

    // Track active connection
    this.activeConnections.add(connectionId);

    // Update metrics
    this.updateMetrics();

    // Check for suspicious activity
    this.checkSuspiciousActivity(ip, connectionData);

    return connectionId;
  }

  public updateConnectionBytes(connectionId: string, bytes: number): void {
    // Find and update the connection
    for (const [ip, connections] of this.connections.entries()) {
      const connection = connections.find(c => 
        this.generateConnectionId(c) === connectionId
      );
      if (connection) {
        connection.bytesTransferred = bytes;
        this.updateMetrics();
        
        // Check for large download
        if (bytes > this.suspiciousThresholds.largeDownload) {
          this.flagSuspiciousActivity(ip, 'large_download', 
            `Large download detected: ${this.formatBytes(bytes)}`, 'medium');
        }
        break;
      }
    }
  }

  public closeConnection(connectionId: string): void {
    this.activeConnections.delete(connectionId);
    
    // Find and update connection duration
    for (const connections of this.connections.values()) {
      const connection = connections.find(c => 
        this.generateConnectionId(c) === connectionId
      );
      if (connection) {
        connection.duration = Date.now() - connection.timestamp;
        break;
      }
    }
    
    this.updateMetrics();
  }

  private generateConnectionId(connection?: ConnectionData): string {
    if (connection) {
      return `${connection.ip}_${connection.timestamp}_${connection.type}`;
    }
    return `${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  private updateMetrics(): void {
    const allConnections: ConnectionData[] = [];
    const uniqueIPs = new Set<string>();

    // Aggregate all connections
    for (const [ip, connections] of this.connections.entries()) {
      uniqueIPs.add(ip);
      allConnections.push(...connections);
    }

    // Calculate metrics
    this.metrics.totalConnections = allConnections.length;
    this.metrics.activeConnections = this.activeConnections.size;
    this.metrics.uniqueIPs = uniqueIPs.size;

    // Connection types
    this.metrics.connectionsByType = {};
    for (const connection of allConnections) {
      this.metrics.connectionsByType[connection.type] = 
        (this.metrics.connectionsByType[connection.type] || 0) + 1;
    }

    // Average duration and total bytes
    const completedConnections = allConnections.filter(c => c.duration);
    this.metrics.averageConnectionDuration = completedConnections.length > 0
      ? completedConnections.reduce((sum, c) => sum + (c.duration || 0), 0) / completedConnections.length
      : 0;

    this.metrics.totalBytesTransferred = allConnections.reduce(
      (sum, c) => sum + (c.bytesTransferred || 0), 0
    );
  }

  private checkSuspiciousActivity(ip: string, newConnection: ConnectionData): void {
    const ipConnections = this.connections.get(ip) || [];
    const recentConnections = ipConnections.filter(
      c => Date.now() - c.timestamp < 60000 // Last minute
    );

    // Check for rapid requests
    if (recentConnections.length >= this.suspiciousThresholds.rapidRequests) {
      this.flagSuspiciousActivity(ip, 'rapid_requests',
        `${recentConnections.length} requests in the last minute`, 'high');
    }

    // Check for unusual patterns
    const typeCount = recentConnections.filter(c => c.type === newConnection.type).length;
    if (typeCount > 5 && newConnection.type === 'file_download') {
      this.flagSuspiciousActivity(ip, 'unusual_pattern',
        `Repeated ${newConnection.type} requests`, 'medium');
    }
  }

  private flagSuspiciousActivity(
    ip: string, 
    type: SuspiciousActivity['type'], 
    details: string, 
    severity: SuspiciousActivity['severity']
  ): void {
    const activity: SuspiciousActivity = {
      ip,
      type,
      timestamp: Date.now(),
      details,
      severity
    };

    this.metrics.suspiciousActivity.push(activity);
    
    // Limit suspicious activity history
    if (this.metrics.suspiciousActivity.length > 100) {
      this.metrics.suspiciousActivity.shift();
    }

    // Emit event for real-time monitoring
    this.emit('suspiciousActivity', activity);
  }

  private startAnalysis(): void {
    this.analysisInterval = setInterval(() => {
      this.cleanupOldConnections();
      this.performSecurityAnalysis();
    }, 30000); // Run every 30 seconds
  }

  private cleanupOldConnections(): void {
    const cutoffTime = Date.now() - (24 * 60 * 60 * 1000); // 24 hours ago
    
    for (const [ip, connections] of this.connections.entries()) {
      const filteredConnections = connections.filter(c => c.timestamp > cutoffTime);
      if (filteredConnections.length === 0) {
        this.connections.delete(ip);
      } else {
        this.connections.set(ip, filteredConnections);
      }
    }

    // Clean up old suspicious activities
    this.metrics.suspiciousActivity = this.metrics.suspiciousActivity.filter(
      activity => Date.now() - activity.timestamp < (24 * 60 * 60 * 1000)
    );
  }

  private performSecurityAnalysis(): void {
    // Analyze connection patterns for potential attacks
    for (const [ip, connections] of this.connections.entries()) {
      const recentConnections = connections.filter(
        c => Date.now() - c.timestamp < 300000 // Last 5 minutes
      );

      // Check for potential DDoS patterns
      if (recentConnections.length > 20) {
        this.flagSuspiciousActivity(ip, 'unusual_pattern',
          'Potential DDoS pattern detected', 'high');
      }

      // Check for data exfiltration patterns
      const totalBytes = recentConnections.reduce(
        (sum, c) => sum + (c.bytesTransferred || 0), 0
      );
      if (totalBytes > 500 * 1024 * 1024) { // 500MB
        this.flagSuspiciousActivity(ip, 'large_download',
          'Potential data exfiltration detected', 'high');
      }
    }
  }

  private formatBytes(bytes: number): string {
    if (bytes >= 1024 * 1024 * 1024) {
      return `${(bytes / (1024 * 1024 * 1024)).toFixed(2)} GB`;
    } else if (bytes >= 1024 * 1024) {
      return `${(bytes / (1024 * 1024)).toFixed(2)} MB`;
    } else if (bytes >= 1024) {
      return `${(bytes / 1024).toFixed(2)} KB`;
    }
    return `${bytes} B`;
  }

  public getMetrics(): ConnectionMetrics {
    return { ...this.metrics };
  }

  public getConnectionHistory(ip?: string): ConnectionData[] {
    if (ip) {
      return this.connections.get(ip) || [];
    }
    
    const allConnections: ConnectionData[] = [];
    for (const connections of this.connections.values()) {
      allConnections.push(...connections);
    }
    return allConnections.sort((a, b) => b.timestamp - a.timestamp);
  }

  public destroy(): void {
    if (this.analysisInterval) {
      clearInterval(this.analysisInterval);
      this.analysisInterval = null;
    }
    this.connections.clear();
    this.activeConnections.clear();
    this.removeAllListeners();
  }
}